### PR TITLE
Added UPS integration

### PR DESCRIPTION
**Pull request by ziccardi**
# Motivation
Issue [AEROGEAR-8488](https://issues.jboss.org/browse/AEROGEAR-8488)

# Description
Added UPS integration to the `ionic-showcase`.
Now, if the app is closed when a change is done on the server, a notification is received on the client.
Clicking on the notification, will open the app and will move to the tasks page.

- Added UPS integration
- Renamed package to `org.aerogear.ionic.showcase`
- Created app in Firebase with package `org.aerogear.ionic.showcase`
- Created app `ionic-showcase` with android variant in the community cluster UPS
- Added a simple `settings` page that allow to enable/disable push notifications

# How to run

1. Add `google-services.json` file into the root of the project
2. add push-config.json into the `server/src/config folder`
3. add push-service configuration to the `mobile-services.js` file. An example:
```json
"services": [
    {
      "id": "push",
      "name": "push",
      "type": "push",
      "url": "https://ups-collab-project.comm2.skunkhenry.com",
      "config": {
        "android": {
          "senderID": "your sender id",
          "variantId": "your variant id",
          "variantSecret": "your variant secret"
        }
      }
    }
  ]
```

### CODE REVIEW

**Comment by wtrocki** [comment]
@ziccardi Awesome work so far!

**Comment by wtrocki** [comment]
@ziccardi I think we could avoid handling push notification when application is in foreground and show notification when in background. If we can start application on task page when launched from notification then it will look cool however it is not critical to do it this way.

File: src/app/app.component.ts [hunk 1]
```diff
@@ -17,7 +18,8 @@ export class AppComponent {
     private platform: Platform,
     private splashScreen: SplashScreen,
     private statusBar: StatusBar,
-    public auth: AuthService
+    public auth: AuthService,
+    private pushService: PushService
```
Comment (wtrocki) [inline src/app/app.component.ts]: Not used

File: src/app/pages/home/home.page.ts [hunk 2]
```diff
@@ -1,10 +1,42 @@
...
 export class HomePage {
+    private background = false;
+
+    constructor(
+        private localNotifications: LocalNotifications,
+        plt: Platform,
+        push: PushService,
+        public navCtrl: NavController) {
+        // We need to wait for the platform to initialize the plugins
+        plt.ready().then(() => {
```
Comment (wtrocki) [inline src/app/pages/home/home.page.ts]: I believe that home page may not be best place to put this. It works, but it is not ideal in terms of showcasing this in code. Home page is one of the many possible places to land in the app. The safest way is to extract this code into the service and call:

`pushService.init()` from `app.component`

File: src/app/pages/home/home.page.ts [hunk 3]
```diff
@@ -1,10 +1,42 @@
...
+            push.initPush();
+            push.setCallback(this.handleNotification.bind(this));
+            push.register();
+
+            if (document) {
+              document.addEventListener('deviceready', () => {
```
Comment (wtrocki) [inline src/app/pages/home/home.page.ts]: What is happening here?

File: src/app/pages/home/home.page.ts [hunk 4]
```diff
@@ -1,10 +1,42 @@
...
+        plt: Platform,
+        push: PushService,
+        public navCtrl: NavController) {
+        // We need to wait for the platform to initialize the plugins
+        plt.ready().then(() => {
+            push.initPush();
+            push.setCallback(this.handleNotification.bind(this));
+            push.register();
+
+            if (document) {
```
Comment (wtrocki) [inline src/app/pages/home/home.page.ts]: Same as above. Really important code tha should be part of the service. 
When documenting showcase we could just point to service for all business logic related to push

File: src/app/services/push.service.ts [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+ */
+@Injectable()
+export class PushService {
+  public static registered = false;
+
+  // We want one single instance & callback app wide
+  public static pushObject: PushObject = null;
+  public static callback: (notification: PushMessage) => void;
+
+  public messages: PushMessage[] = [];
```
Comment (wtrocki) [inline src/app/services/push.service.ts]: Unused. Let's remove

File: src/app/services/push.service.ts [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+    });
+
+    PushService.pushObject.on('notification').subscribe((notification) => {
+      const newNotification = {
+        title: notification.title,
+        message: notification.message,
+        received: new Date().toDateString(),
+      };
+      this.messages.push(newNotification);
+      this.emit(newNotification);
```
Comment (wtrocki) [inline src/app/services/push.service.ts]: Is that used.

**Comment by wtrocki** [comment]
@ziccardi I have approved that prematurely. Going to verify changes now

File: server/src/tasks/schema.js [hunk 7]
```diff
@@ -95,8 +97,16 @@ const taskResolvers = {
 function publish(actionType, data, pushClient) {
   if (pushClient) {
     pushClient.sender.send({
-      alert: `${actionType} ${data.title}`
-    }).then((response) => {
+      userData: {
+        title: data.title,
```
Comment (StephenCoady) [inline server/src/tasks/schema.js]: should there be any string parsing here? excuse my lack of knowledge about push but what happens if we send a really long and ugly item? will the notification automatically add ellipses? 

So will `Thisisareallylongtexttodotitle` automatically become `Thisisareally....` ?

File: src/app/services/push.service.ts [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+        console.log('Push registration successful');
+      }).catch((err) => {
+        this.pushError = err;
+        console.error('Push registration unsuccessful ', this.pushError);
+      });
+    });
+
+    PushService.pushObject.on('notification').subscribe((notification) => {
+      const newNotification = {
+        title: notification.title,
```
Comment (StephenCoady) [inline src/app/services/push.service.ts]: this is probably a more appropriate place to ask about the string parsing. we should probably do it here.

File: src/app/pages/home/home.page.ts (hunk 3 above)
Comment (ziccardi) [reply src/app/pages/home/home.page.ts]: It is code to detect if the app is in foreground or in background. Used to ignore the notification when receiving it and the app is in foreground.

File: src/app/pages/home/home.page.ts (hunk 4 above)
Comment (ziccardi) [reply src/app/pages/home/home.page.ts]: 👌

File: server/src/tasks/schema.js (hunk 7 above)
Comment (StephenCoady) [reply server/src/tasks/schema.js]: I can answer this myself now after investigating. The notifications tool does parsing for us so ignore me 👍

File: src/app/pages/home/home.page.ts (hunk 3 above)
Comment (wtrocki) [reply src/app/pages/home/home.page.ts]: Nice!

**Comment by wtrocki** [comment]
@ziccardi Do we have screen for enabling disabling UPS? It works great but I think that after talking with @wei-lee we have asked for Settings page when push can be enabled disabled.

**Comment by ziccardi** [comment]
@wtrocki : going to add it, I don't think it will be too hard.

File: src/app/services/push.service.ts (hunk 6 above)
Comment (ziccardi) [reply src/app/services/push.service.ts]: yes, it is used to invoke the callback when a notification is received.

File: src/app/services/push.service.ts (hunk 5 above)
Comment (ziccardi) [reply src/app/services/push.service.ts]: 👍

**Pull request merged**